string termination
float32 reward